### ISSUE TITLE
Question about sequence-to-sequence network

### ISSUE THREAD

**Issue by tmfarrell** [Element-Research/rnn#240]
Hi everyone, 

I'm trying to build a sequencer-to-sequence RNN by extending the `encoder-decoder-coupling.lua` example.  And can't seem to figure out an "index out of range" error I keep getting. 

I'm new to RNNs and torch so assume it's probably an issue of not knowing how to initialize the encoder/ decoders.  

Any help appreciated. 

Here's the code: 
``` lua
require 'rnn'
require 'hdf5'

-- Get Data 
print('Reading data...') 
h5 = hdf5.open('../data/data.h5', 'r')
-- input signals to the encoder
local encInSeq = h5:read('Data/Signals'):all()
-- input sequences to decoder 
local decInSeq = h5:read('Data/SequencesIn'):all()
-- expected output sequences 
local decOutSeq = h5:read('Data/SequencesOut'):all()
-- decoder predicts one per timestep, so split accordingly
decOutSeq = nn.SplitTable(1, 1):forward(decOutSeq)
print('Number seqs: ' .. encInSeq:size(1)) 
print('Length of encoder input seqs: ' .. encInSeq:size(2)) 
print('Length of decoder input seqs: ' .. decInSeq:size(2)) 
print('decOutSeq:')
print (decOutSeq) 

print('Setting hyperparameters...') 
version = 1.2 
local opt = {}
opt.niter = 50
opt.learningRate = 0.1
opt.seqLen = decInSeq:size(2) 
opt.signalLen = encInSeq:size(2) 
opt.hiddenSize = 1024   -- # possible characters in decInSeq/ decOutSeq

print('Initing forward and backward LSTM connect functions...') 
--[[ Forward coupling: Copy encoder cell and output to decoder LSTM ]]--
local function forwardConnect(encLSTM, decLSTM)
   decLSTM.userPrevOutput = nn.rnn.recursiveCopy(decLSTM.userPrevOutput, encLSTM.outputs[opt.seqLen])
   decLSTM.userPrevCell = nn.rnn.recursiveCopy(decLSTM.userPrevCell, encLSTM.cells[opt.seqLen])
end
--[[ Backward coupling: Copy decoder gradients to encoder LSTM ]]--
local function backwardConnect(encLSTM, decLSTM)
   encLSTM.userNextGradCell = nn.rnn.recursiveCopy(encLSTM.userNextGradCell, decLSTM.userGradPrevCell)
   encLSTM.gradPrevOutput = nn.rnn.recursiveCopy(encLSTM.gradPrevOutput, decLSTM.userGradPrevOutput)
end

-- Encoder
print('Building encoder...') 
local enc = nn.Sequential() 
enc:add(nn.LookupTable(opt.signalLen, opt.hiddenSize))  -- encode in signal -> hidden
enc:add(nn.SplitTable(1, 2)) -- works for both online and mini-batch mode
local encLSTM = nn.LSTM(opt.hiddenSize, opt.seqLen)    -- hidden -> decode in seq
enc:add(nn.Sequencer(encLSTM))
enc:add(nn.SelectTable(-1))
print(enc) 

-- Decoder
print('Building decoder...') 
local dec = nn.Sequential()
dec:add(nn.LookupTable(opt.seqLen, opt.hiddenSize))  -- decode in seq -> hidden
dec:add(nn.SplitTable(1, 2)) -- works for both online and mini-batch mode
local decLSTM = nn.LSTM(opt.hiddenSize, opt.hiddenSize)  -- hidden -> hidden 
dec:add(nn.Sequencer(decLSTM))
dec:add(nn.Sequencer(nn.Linear(opt.hiddenSize, opt.seqLen)))  -- hidden -> decode out seq
dec:add(nn.Sequencer(nn.LogSoftMax()))
print(dec) 

local criterion = nn.SequencerCriterion(nn.ClassNLLCriterion())

print('Training...') 
for i=1, opt.niter do
   enc:zeroGradParameters()
   dec:zeroGradParameters()

   -- Forward pass
   local encOut = enc:forward(encInSeq)  -- fails here (I think), with "index out of range" 
   forwardConnect(encLSTM, decLSTM)
   local decOut = dec:forward(decInSeq)
   local err = criterion:forward(decOut, decOutSeq)
   
   print(string.format("Iteration %d ; NLL err = %f ", i, err))

   -- Backward pass
   local gradOutput = criterion:backward(decOut, decOutSeq)
   dec:backward(decInSeq, gradOutput)
   backwardConnect(encLSTM, decLSTM)
   local zeroTensor = torch.Tensor(encOut):zero()
   enc:backward(encInSeq, zeroTensor)

   dec:updateParameters(opt.learningRate)
   enc:updateParameters(opt.learningRate)
end
```

The output:
```
Importing modules...	
Reading data...	
Number seqs: 5	
Length of encoder input seqs: 32815	
Length of decoder input seqs: 79	
decOutSeq:	
{
  1 : FloatTensor - size: 5
  2 : FloatTensor - size: 5
  3 : FloatTensor - size: 5
  4 : FloatTensor - size: 5
  5 : FloatTensor - size: 5
  6 : FloatTensor - size: 5
  7 : FloatTensor - size: 5
  8 : FloatTensor - size: 5
  9 : FloatTensor - size: 5
  10 : FloatTensor - size: 5
  11 : FloatTensor - size: 5
  12 : FloatTensor - size: 5
  13 : FloatTensor - size: 5
  14 : FloatTensor - size: 5
  15 : FloatTensor - size: 5
  16 : FloatTensor - size: 5
  17 : FloatTensor - size: 5
  18 : FloatTensor - size: 5
  19 : FloatTensor - size: 5
  20 : FloatTensor - size: 5
  21 : FloatTensor - size: 5
  22 : FloatTensor - size: 5
  23 : FloatTensor - size: 5
  24 : FloatTensor - size: 5
  25 : FloatTensor - size: 5
  26 : FloatTensor - size: 5
  27 : FloatTensor - size: 5
  28 : FloatTensor - size: 5
  29 : FloatTensor - size: 5
  30 : FloatTensor - size: 5
  31 : FloatTensor - size: 5
  32 : FloatTensor - size: 5
  33 : FloatTensor - size: 5
  34 : FloatTensor - size: 5
  35 : FloatTensor - size: 5
  36 : FloatTensor - size: 5
  37 : FloatTensor - size: 5
  38 : FloatTensor - size: 5
  39 : FloatTensor - size: 5
  40 : FloatTensor - size: 5
  41 : FloatTensor - size: 5
  42 : FloatTensor - size: 5
  43 : FloatTensor - size: 5
  44 : FloatTensor - size: 5
  45 : FloatTensor - size: 5
  46 : FloatTensor - size: 5
  47 : FloatTensor - size: 5
  48 : FloatTensor - size: 5
  49 : FloatTensor - size: 5
  50 : FloatTensor - size: 5
  51 : FloatTensor - size: 5
  52 : FloatTensor - size: 5
  53 : FloatTensor - size: 5
  54 : FloatTensor - size: 5
  55 : FloatTensor - size: 5
  56 : FloatTensor - size: 5
  57 : FloatTensor - size: 5
  58 : FloatTensor - size: 5
  59 : FloatTensor - size: 5
  60 : FloatTensor - size: 5
  61 : FloatTensor - size: 5
  62 : FloatTensor - size: 5
  63 : FloatTensor - size: 5
  64 : FloatTensor - size: 5
  65 : FloatTensor - size: 5
  66 : FloatTensor - size: 5
  67 : FloatTensor - size: 5
  68 : FloatTensor - size: 5
  69 : FloatTensor - size: 5
  70 : FloatTensor - size: 5
  71 : FloatTensor - size: 5
  72 : FloatTensor - size: 5
  73 : FloatTensor - size: 5
  74 : FloatTensor - size: 5
  75 : FloatTensor - size: 5
  76 : FloatTensor - size: 5
  77 : FloatTensor - size: 5
  78 : FloatTensor - size: 5
  79 : FloatTensor - size: 5
}
Setting hyperparameters...	
Initing forward and backward LSTM functions...	
Building encoder...	
nn.Sequential {
  [input -> (1) -> (2) -> (3) -> (4) -> output]
  (1): nn.LookupTable
  (2): nn.SplitTable
  (3): nn.Sequencer @ nn.LSTM(1024 -> 79)
  (4): nn.SelectTable
}
Building decoder...	
nn.Sequential {
  [input -> (1) -> (2) -> (3) -> (4) -> (5) -> output]
  (1): nn.LookupTable
  (2): nn.SplitTable
  (3): nn.Sequencer @ nn.LSTM(1024 -> 1024)
  (4): nn.Sequencer @ nn.Recursor @ nn.Linear(1024 -> 79)
  (5): nn.Sequencer @ nn.Recursor @ nn.LogSoftMax
}
Training...	
```

And the error:
```
/share/pkg/torch/f972c42/install/bin/luajit: ...pkg/torch/f972c42/install/share/lua/5.1/nn/Container.lua:69: 
In 1 module of nn.Sequential:
...g/torch/f972c42/install/share/lua/5.1/nn/LookupTable.lua:62: index out of range at /share/pkg/torch/f972c42/src/pkg/torch/lib/TH/generic/THTensorMath.c:156
stack traceback:
	[C]: in function 'index'
	...g/torch/f972c42/install/share/lua/5.1/nn/LookupTable.lua:62: in function <...g/torch/f972c42/install/share/lua/5.1/nn/LookupTable.lua:56>
	[C]: in function 'xpcall'
	...pkg/torch/f972c42/install/share/lua/5.1/nn/Container.lua:65: in function 'rethrowErrors'
	...kg/torch/f972c42/install/share/lua/5.1/nn/Sequential.lua:44: in function 'forward'
	ont_encoder_decoder_rnn.lua:88: in main chunk
	[C]: in function 'dofile'
	...ch/f972c42/install/lib/luarocks/rocks/trepl/scm-1/bin/th:145: in main chunk
	[C]: at 0x00405800

WARNING: If you see a stack trace below, it doesn't point to the place where this error occured. Please use only the one above.
stack traceback:
	[C]: in function 'error'
	...pkg/torch/f972c42/install/share/lua/5.1/nn/Container.lua:69: in function 'rethrowErrors'
	...kg/torch/f972c42/install/share/lua/5.1/nn/Sequential.lua:44: in function 'forward'
	ont_encoder_decoder_rnn.lua:88: in main chunk
	[C]: in function 'dofile'
	...ch/f972c42/install/lib/luarocks/rocks/trepl/scm-1/bin/th:145: in main chunk
	[C]: at 0x00405800
```

**Comment by rahul-iisc** [comment]
@tmfarrell The first argument of the LookupTable should be size of vocabulary. Please make sure that is correct. Also, you have different first arguments in LookupTables of the encoder and the decoder, please check if vocab-size is really different for your source and target sequences.

**Comment by khushigupta** [comment]
Also make sure that the decOutSeq has 1 instead of 0 padding, to avoid a similar error.

**Comment by tmfarrell** [comment]
@rahul-iisc the source and target seqs do have different vocabs sizes, but based on your comment I'll have to adjust those numbers. Thanks for the suggestion, I'll let you know how it goes

@khushigupta I did 0-pad. I'll switch to 1-padding. Thanks as well
